lib/bing-ads-reporting/client.rb

```diff
@@ -0,0 +1,86 @@
+module BingAdsReporting
+  class Client
+    include Formatter
+    API_CALL_RETRY_COUNT = 3
+
+    def initialize(settings, logger)
+      soap_header = header(settings)
+      log_level = settings[:log_level] || :info
+      @logger = logger
+      @soap_client = Savon.client({
+        wsdl: "https://api.bingads.microsoft.com/Api/Advertiser/Reporting/V9/ReportingService.svc?wsdl",
+        namespaces: {"xmlns:arr" => 'http://schemas.microsoft.com/2003/10/Serialization/Arrays'},
+        soap_header: soap_header,
+        log_level: log_level,
+        pretty_print_xml: true
+      })
+    end
+
+    def call(service, message, retry_count = API_CALL_RETRY_COUNT)
+      1.upto(retry_count + 1) do |retry_index|
+        begin
+          response = @soap_client.call(service, message: message)
+          break response
+        rescue Savon::SOAPFault => error
+          next if retry_index <= retry_count
+          break handle_soap_fault(error)
+        end
+      end
+    end
+
+    def download(url, retry_count = API_CALL_RETRY_COUNT)
+      1.upto(retry_count + 1) do |retry_index|
+        begin
+          @logger.info "Downloading Bing material from: #{url}"
+          curl = Curl::Easy.new(url)
+          curl.perform
+          body = curl.body_str
+          break body
+        rescue => ex
+          next if retry_index <= retry_count
+          raise DownloadError, ex.message
+        end
+      end
+    end
+
+    private
+
+    def header(settings)
+      if settings[:username] && settings[:password]
+        {
```

Comment:
the 2 branches are identical, except the `username` and `password`. why repeat the rest?